src/python/pants/engine/internals/defaults.py

```diff
@@ -0,0 +1,201 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+from dataclasses import dataclass, field, replace
+from typing import Any, Iterable, Mapping, Tuple, Union
+
+from pants.engine.addresses import Address
+from pants.engine.target import InvalidFieldException, RegisteredTargetTypes
+from pants.engine.unions import UnionMembership
+from pants.util.frozendict import FrozenDict
+
+DefaultFieldValuesT = FrozenDict[str, Any]
+DefaultsValueT = FrozenDict[str, DefaultFieldValuesT]
+DefaultsT = FrozenDict[str, DefaultsValueT]
+
+SetDefaultsValueT = Mapping[str, Any]
+SetDefaultsKeyT = Union[str, Tuple[str, ...]]
+SetDefaultsT = Mapping[SetDefaultsKeyT, SetDefaultsValueT]
+
+
+@dataclass
+class BuildFileDefaultsProvider:
```

Comment:
So, rather than mutability, you should be able to use upward recursion in the filesystem, which can be memoized. Basically: the defaults for directory `x/y/z` are (recursively) the defaults for `z/y` + the defaults for `z`.

An example where we do this is source root stripping: `get_optional_source_roots` is self recursive: https://github.com/pantsbuild/pants/blob/04cc7b4bb31066f5020b5d2fe9ff623ddc63dd2d/src/python/pants/source/source_root.py#L210-L235

Reply:
Can probably remove this class now, and inline the two fields into `@rules` that need them, with `get_parser_defaults` as a factory function of `BuildFileDefaultsParserState`.